### PR TITLE
No longer double count transfer cost in stealing

### LINKED ISSUE #7002: Stealing balance not accounting for recent decisions
The [Stealing.balance](https://github.com/dask/distributed/blob/b133009cee88fd48c8a345cffde0a8e9163426a6/distributed/stealing.py#L378-L502) implementation is currently using a greedy for-loop with multiple early termination conditions. These break statements are de-facto dead code causing too greedy and potentially destructive work stealing.

Stealing is highly sensitive to the definition of idle and saturated workers which is defined in [`Scheduler.check_idle_saturated`](https://github.com/dask/distributed/blob/b133009cee88fd48c8a345cffde0a8e9163426a6/distributed/scheduler.py#L2812-L2863).

<details>

<summary> Explanation of saturated </summary>

See also https://github.com/dask/distributed/pull/6614/files/36a60a5e358ea2a5d16597651126ac5892203b01#r952608704 

I think the new definition of idle to incorporate whether there are free slots makes a lot of sense.

> I didn't change the saturated definition because I don't honestly understand what it's trying to measure: https://github.com/dask/distributed/blob/a1d2011e7d5d5ba26bea8bd4b0dec65f485f2b43/distributed/scheduler.py#L2611-L2613

off-topic but I spent some time on this before and it's worth sharing

you can rewrite this statement to

```python
pending: float = occ * (p - nc) / (p * nc)
pending = occ / nc - occ / p
pending = occupancy / n_threads - occupancy / processing
pending = avg_occ_per_thread - avg_occ_currently_processing
```

i.e. `pending` is "average occupancy per thread that is in state `ready` / not, yet executing"
or put into context of stealing, which is the only place that `saturated` matters, "occupancy per thread that could be stolen"
I guess what's written there is more stable in terms of floating point arithmetic.

So, the definition of saturated is spelled out something like

A worker is classified as saturated if it's `ready` occupancy is at least 0.4s per thread and at least 1.9 times the average cluster wide occupancy.

i.e. "there is something queued up / in ready and it is about two times more than average"

</details>

Simplified, this is how the algorithm is _intended_ to work:
`Stealing.balance` iterates over _saturated_ workers in descending occupancy and victimizes them for stealing. An _idle_ worker will act as the thief and work will be rebalanced. This continues until there are no longer saturated workers or there are no longer idle workers.

However, this pseudo algorithm is no longer implemented. The current code only shows hints at how this was formerly achieved and it likely strongly depended on dicts, sets, etc. being passed by reference such that any changes to the fundamental collections are immediately propagated to the balance loop which then allows for early termination. 

`Stealing.maybe_move_task ` is calling `Scheduler. check_idle_saturated` with an additional keyword `occ` which is *only used here*. This keyword allows to _override_ the actual occupancy of the worker which would allow us to reclassify thieves and victims based on their `combined_occupancy`, i.e. the `Scheduler.idle` and `Scheduler.saturated` collections are updated under the assumption that in-flight tasks are stolen successfully. https://github.com/dask/distributed/blob/b133009cee88fd48c8a345cffde0a8e9163426a6/distributed/stealing.py#L409-L410

Assuming that `idle` in the balance for-loop was referencing `Scheduler.idle` and not be a derivative, this would allow us to break out of the loop at various places, e.g. [here](https://github.com/dask/distributed/blob/b133009cee88fd48c8a345cffde0a8e9163426a6/distributed/stealing.py#L436-L437), [here](https://github.com/dask/distributed/blob/b133009cee88fd48c8a345cffde0a8e9163426a6/distributed/stealing.py#L451-L452) or [here](https://github.com/dask/distributed/blob/b133009cee88fd48c8a345cffde0a8e9163426a6/distributed/stealing.py#L471-L472). However, the loop variable `idle` is *always* [defined as a list](https://github.com/dask/distributed/blob/b133009cee88fd48c8a345cffde0a8e9163426a6/distributed/stealing.py#L415) since https://github.com/dask/distributed/pull/5665 and even before [is _sometimes_ a list due to sorting](https://github.com/dask/distributed/blob/b133009cee88fd48c8a345cffde0a8e9163426a6/distributed/stealing.py#L432-L433) ever since https://github.com/dask/distributed/pull/754
This effectively renders any `if not idle; break` statements dead code. This can be confirmed by codecoverage measurements

image

Going back to https://github.com/dask/distributed/pull/750 it appears that a similar mechanism existed for early termination if saturated became empty. The implementation changed drastically and I do not see any leftover termination mechanism that ensures a potential _victim_ is still saturated if one accounts for combined_occupancy.

I suspect this lack of early termination to be a major contributor to at least the following tickets

- https://github.com/dask/distributed/issues/6573
- https://github.com/dask/distributed/issues/5243

### PR DESCRIPTION
**Pull request by hendrikmakait**
* Closes #7002
* Supersedes #7026

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/stealing.py [hunk 1]
```diff
@@ -400,119 +397,99 @@ def balance(self) -> None:
...
                 ]
-            elif len(saturated) < 20:
-                saturated = sorted(saturated, key=combined_occupancy, reverse=True)
-            if len(idle) < 20:
-                idle = sorted(idle, key=combined_occupancy)
-
-            for level, cost_multiplier in enumerate(self.cost_multipliers):
-                if not idle:
+                if not potential_victims:
+                    # TODO: Unclear how to reach this and what the implications
```
Comment (hendrikmakait) [inline distributed/stealing.py:420]: @fjetter: Is there something we need to do here?

**Review by fjetter — APPROVED** [review]

File: distributed/stealing.py [hunk 2]
```diff
@@ -234,14 +242,13 @@ def steal_time_ratio(self, ts: TaskState) -> tuple[float, int] | tuple[None, Non
...
-        ws = ts.processing_on
-        compute_time = ws.processing[ts]
+        compute_time = self.scheduler.get_task_duration(ts)
 
         if not compute_time:
             # occupancy/ws.proccessing[ts] is only allowed to be zero for
             # long running tasks which cannot be stolen
-            assert ts in ws.long_running
+            assert ts.processing_on
+            assert ts in ts.processing_on.long_running
```
Comment (fjetter) [inline distributed/stealing.py:251]: Codecov says this is not hit but I'd expect this to be hit by `test_dont_steal_long_running_tasks`

We should probably look at that test, maybe some test assumption is no longer true

Comment (fjetter) [reply distributed/stealing.py:251]: Ok, `test_dont_steal_long_running_tasks` is pretty meaningless. I believe something went wrong during a refactoring of that test. It only asserts that long_tasks are only "executed" once, i.e. they are not rescheduled. This is entirely irrelevant for stealing.

Comment (fjetter) [reply distributed/stealing.py:251]: Given that this test was already this way I suggest to set this out of scope for this PR. I might be able to address this with https://github.com/dask/distributed/pull/7030

**Comment by fjetter** [comment]
All failing tests are known offenders except of `test_wait_first_completed` but I opened https://github.com/dask/distributed/pull/7039 to make that one more robust

**Comment by hendrikmakait** [comment]
A/B test: https://github.com/coiled/coiled-runtime/actions/runs/3067074093

**Pull request merged**